Checkpoint/11 05 25

hariharan-devarajan opened the pull request.

hariharan-devarajan committed:
merge commit for checkpoint 10-29-25

hariharan-devarajan committed:
Fixes separate install user vs run user for log and trace file creation.
Fixes #124

hariharan-devarajan committed:
Streamlined the Argparsing for executables

hariharan-devarajan committed:
Fixes Setup datacrumbs by setting limits
Fixes #123

hariharan-devarajan committed:
Fixes Make client so patch for uprobe point to cmake install prefix
Fixes #122

hariharan-devarajan committed:
Fixes for installing datacrumbs

1. Cleaned folder structure for build and install
2. #127 Create datacrumbs script to patch
libdatacrumbs_client into executable
3. #128 Write datacrumbs wrapper to run the app
4. #129 Load datacrumbs skeleton as well in
validator to validate probes
5. Datacrumbs should be installed in sbin not bin
Fixes #130
6. #133 Datacrumbs system deployment
7. datacrumbs_setup need execution flag.
Fixes #134

hariharan-devarajan committed:
Support for Composable Datacrumbs

- Install So and required h and cpp files for building composites.
- Build infrastructure around this
- multi configuration datacrumbs
Fixes #121

hariharan-devarajan committed:
Write a artifact file given by user which contains high level info about run like
Fixes #138

hariharan-devarajan committed:
Added feature for arg parsing and logging

1. #141 Set all variables as readonly
2. Improve logging functions to include new lines and better formatting
3. Add dry-run support to datacrumbs_compose_run script

hariharan-devarajan committed:
Refactor EventProcessor and Server Implementation

- Moved the implementation of the EventProcessor class from event_processor.cpp to a new file server.impl.cpp for better organization.
- Updated server.cpp to include the new server.impl.cpp and modified the main function to call main_call.
- Added a finalize method declaration in event_processor.h, removing the previous implementation from the header file.
- Enhanced the ChromeWriter class by cleaning up unnecessary includes and comments in both chrome_writer.cpp and chrome_writer.h.
- Improved logging and error handling in various functions to ensure better traceability and debugging capabilities.

hariharan-devarajan committed:
feat: Add MPI support and enhance job configuration

- Introduced MPI support in the configuration manager, including rank and size tracking.
- Updated job launch logic to conditionally enable MPI based on user input.
- Enhanced argument parsing to include options for enabling MPI, setting the number of nodes, and processes per node.
- Modified server and job launch functions to handle MPI configurations.
- Improved logging to include rank-specific information for better traceability in MPI environments.
- Adjusted file naming conventions to incorporate MPI rank and size for generated files.
- Ensured that only the root rank logs certain messages to reduce clutter in multi-process scenarios.

hariharan-devarajan committed:
adding flux integration

1. Built a server start and stop script
2.  Built a service for systemctl
3. added service run and service start scripts for flux job prolog and epilog.

milroy commented on src/datacrumbs/server/process/event_processor.cpp:
```diff
@@ -871,25 +871,20 @@ int main_process(int argc, char** argv, datacrumbs::EventProcessor* event_proces
+  auto event_processor = datacrumbs::EventProcessor(argc, argv);
   std::string hostname = get_hostname();
   std::string timestamp = get_timestamp();
   std::string pidfile = "/tmp/datacrumbs_" + hostname + ".pid";
```
This should be in `PREFIX_RUN`. Cmake will set it (could be `INSTALL_RUN`).

hariharan-devarajan replied:
DONE

hariharan-devarajan committed:
PR fixes recommendations.

1. We use DATACRUMBS_INSTALL_RUNSTATEDIR from GNUInstallUtils to get the variable.
Make checking of effective user id secure
2. We need to ensure that the effective user id is checked securely to prevent any potential security issues.
3. renamed infrastrutcure to infrastructure
4. set -e pipefail on all scripts.
5. Move build process of composable into sub shell.
6. Updated CMake minimum version to 3.20
7. Change yaml parsing to env file sourcing.
8. Remove enabling of datacrumbs service.

milroy commented on scripts/datacrumbs/datacrumbs_setup.in:
```diff
@@ -0,0 +1,33 @@
+ulimit -n $(ulimit -H -n)
+ulimit -l $(ulimit -H -l)
+
+alias datacrumbs_run="sudo  @CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_SBINDIR@/datacrumbs_run $USER"
```
Stopped here 11/13/2025.

milroy commented on scripts/datacrumbs/datacrumbs_utility.in:
```diff
@@ -0,0 +1,46 @@
+  DATA_USER=$1
+  DATACRUMBS_LOG_PATH=$2
+  if [ ! -d "$DATACRUMBS_LOG_PATH" ]; then
+    mkdir -p "$DATACRUMBS_LOG_PATH"
```
Set permissions on `mkdir` to avoid the subsequent `chmod`s

milroy replied:
The current files changes don't seem to include this fix.

milroy replied:
Clarified in a discussion: the `-eo` option will address this.

milroy commented on scripts/datacrumbs/datacrumbs_compose.in:
```diff
@@ -0,0 +1,96 @@
+    mkdir -p "${BUILD_DIR}"
+    chown "${DATACRUMBS_USER}:${DATACRUMBS_USER}" "${BUILD_DIR}"
+    log "Copying composable files to build directory"
+    cp -r ${DATACRUMBS_INSTALL_COMPOSABLE_DIR}/* ${BUILD_DIR}/
```
Suggestion: build as non-root with subshell.

hariharan-devarajan commented on etc/datacrumbs/configs/project.yaml.in:
convert it into a env file which is directly sourced.

hariharan-devarajan replied:
Done

milroy commented on scripts/datacrumbs/datacrumbs_utility.in:
```diff
@@ -24,18 +24,61 @@ function datacrumbs_finalize() {
-    chmod o-w "$DATACRUMBS_LOG_PATH"
-  fi
+
+  sudo -u ${DATA_USER} mkdir -p ${DATACRUMBS_LOG_PATH}
```
Set permissions in `mkdir` or remove.

hariharan-devarajan replied:
Done

milroy commented on CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.4)
+cmake_minimum_required(VERSION 3.5)
```
Halfway through this file on 12/11.

milroy commented on src/datacrumbs/server/process/server.impl.cpp:
```diff
@@ -0,0 +1,812 @@
+  json_object_put(status_json);
+  pwd = getpwnam(event_processor->configManager_->user.c_str());
+  uid = pwd ? pwd->pw_uid : static_cast<uid_t>(-1);
+  gid = pwd ? pwd->pw_gid : static_cast<gid_t>(-1);
```
@trws, is the `-1` value correct here?

trws replied:
It depends on what this is used for.  I'm pretty sure the "nobody" user on debian-style systems or any system with classic NFS users uses that as its default UID. If this is meant to be a "can't exist" or similar, it's not that.

milroy commented on CMakeLists.txt:
```diff
@@ -324,7 +324,7 @@ foreach(_line ${_CAPTURE_PROBES_NAME_LINES})
+message(STATUS "[${UPPER_PROJECT_NAME}] Capture probes: ${DATACRUMBS_CAPTURE_PROBES_LIST}")
 
 list(APPEND DATACRUMBS_BPF_EXTRA_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/src/datacrumbs/server/bpf/common.c)
 list(APPEND DATACRUMBS_BPF_EXTRA_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/src/datacrumbs/server/bpf/init.bpf.c)
```
Finished this commit 1/15/2026.

hariharan-devarajan commented on src/datacrumbs/composable/CMakeLists.txt.in:
```diff
@@ -0,0 +1,142 @@
+cmake_minimum_required(VERSION 3.4)
```
Done

milroy commented on scripts/datacrumbs/datacrumbs_service_run.sh.in:
```diff
@@ -0,0 +1,10 @@
+    umask 022
+    printenv > /run/datacrumbs.env
+    echo datacrumbs_service_run "Starting datacrumbs service run script on job ${FLUX_JOB_ID}"
+    sudo systemctl enable datacrumbs@$DATACRUMBS_COMPOSABLE_NAME.service
```
This line needs to be removed, because we don't want datacrumbs to start on LC system boot.

milroy commented on notes.MD:
```diff
@@ -6,4 +6,109 @@
+flux run -N $nodes --tasks-per-node 16 ~/software/ior/install/bin/ior -t 1m -b 16m -F -o /tmp/file2
+flux run -N $nodes  --tasks-per-node 16 ~/software/ior/install/bin/ior -a MPIIO -t 1m -b 16m -F -o /tmp/file2
+flux run -N $nodes --tasks-per-node 16 ~/software/ior/install/bin/ior -c -t 1m -b 16m -F -o /tmp/file2 -z
+flux run -N $nodes  --tasks-per-node 16 ~/software/ior/install/bin/ior -t 1m -b 16m -F -o /tmp/file2
```
Stopped here 1/22/2026.

milroy reviewed (commented):
Looks good, except I don't see that permissions are set on `mkdir`. Once that's fixed I will approve.

milroy reviewed (approved):
Approving after discussion on `mkdir`.
